### PR TITLE
Never bind INADDR_ANY by default, and warn when doing so explicitly

### PR DESCRIPTION
**Pull request by luke-jr**
A disturbingly large number of listening nodes appear to be also exposing their RPC server to the public internet. To attempt to mitigate this:

* Only ever bind localhost by default, even if `rpcallowip` is specified. (A warning is given if `rpcallowip` is specified without `rpcbind`, since it doesn't really make sense to do.)
* Warn about exposing the RPC server to untrusted networks if the user explicitly binds to any INADDR_ANY address.
* Include a warning about untrusted networks in the `--help` documentation for `rpcbind`.

### CODE REVIEW

**Comment by laanwj** [comment]
what is the problem with binding to INADDR_ANY if it's rejecting all but specific IPs?

**Comment by kristapsk** [comment]
If we change this default behaviour, it must be mentioned in the release notes.

**Comment by practicalswift** [comment]
Very nice find @luke-jr 

Very strong concept ACK

The expected contract for security or privacy sensitive daemons is to never bind to all interfaces unless the user explicitly asks for that.

@laanwj 

Consider a machine with three IP-addresses on different networks: `A`, `B` and `C`.

Let’s say that all Bitcoin related activity such as outgoing connections etc takes place on network `A`.

**First: the privacy leak aspect**

We should not assume that the user is willing to share the information that he/she is running `bitcoind` also to participants on network `B` or `C`.

The fact that we are not providing any services to unknown clients (thanks to `!ClientAllowed(…)`) does not stop that privacy leak. The privacy is leak happens with the `SYN`/`ACK` response.

**Second: the attack surface aspect**

Even if we are not providing any services to unknown clients (due to `!ClientAllowed(…)`) we’re still providing adversaries on network `B` and `C` an unnecessary attack surface by listening on those networks.

**Comment by luke-jr** [comment]
@laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

correct!

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some trouble it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

correct!

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some trouble it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

Correct! Even though the early stage at which non-allowed connections are dropped makes it unlikely for there to be significant attack surface.

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some craftiness it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

So yes this is a step forward.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

Correct! Though, the early stage at which non-allowed connections are dropped makes it unlikely for there to be significant attack surface.

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some craftiness it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

So yes this is a step forward.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

Correct! Though, the early stage at which non-allowed connections are dropped makes it unlikely for there to be significant attack surface.

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some craftiness it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

So yes this is a step forward.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

Correct! Though, the early stage at which non-allowed connections are dropped makes it unlikely for there to be significant attack surface.

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some craftiness it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

So yes this is a step forward.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

Correct! Though, the early stage at which non-allowed connections are dropped makes it unlikely for there to be significant attack surface.

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some craftiness it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

So yes this is a step forward.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

Correct! Though, the early stage at which non-allowed connections are dropped makes it unlikely for there to be significant attack surface.

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some craftiness it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

So yes this is a step forward.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

Correct! Though, the early stage at which non-allowed connections are dropped makes it unlikely for there to be significant attack surface.

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some craftiness it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

So yes this is a step forward.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

Correct! Though, the early stage at which non-allowed connections are dropped makes it unlikely for there to be significant attack surface.

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some craftiness it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

So yes this is a step forward.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

Correct! Though, the early stage at which non-allowed connections are dropped makes it unlikely for there to be significant attack surface.

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some craftiness it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

So yes this is a step forward.

**Comment by laanwj** [comment]
> @laanwj I don't think we've done anything to make the RPC server robust to connection attacks?

Correct! Though, the early stage at which non-allowed connections are dropped makes it unlikely for there to be significant attack surface.

I love how clightning handles this: RPC is a **local** interface exposed through a UNIX socket only
sure, with some craftiness it's possible to forward it to other machines over ssh, but there's none of that logic in the client

And I think it was a mistake how bitcoind does this, it's one of those historical things, in an ideal world I'd like to get rid of all binding/IP access control logic around RPC. But it seems unrealistic to change any of that, at this stage, that'll result in even more user complaints than this.

So yes this is a step forward.

**Comment by promag** [comment]
Concept ACK.

**Comment by javiertuerca** [comment]
Hello, I am synchronizing a node with the following configuration:

(default: 0 = disable pruning blocks, 1 = allow manual pruning via RPC, greater than 550 = automatically prune blocks to stay under target size in MiB).
prune = 10000

run the test with nmap:

PORT STATE SERVICE

8332 / tcp closed unknown

Is it safe (allow manual pruning via RPC) or should I disable it?

**Comment by laanwj** [comment]
utACK e0520135795563f13feedb28bad5aa3a7e9bc15d

**Comment by MeshCollider** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/14532/commits/e0520135795563f13feedb28bad5aa3a7e9bc15d

File: src/httpserver.cpp [hunk 1]
```diff
@@ -296,9 +296,12 @@ static bool HTTPBindAddresses(struct evhttp* http)
     std::vector<std::pair<std::string, uint16_t> > endpoints;
 
     // Determine what addresses to bind to
-    if (!gArgs.IsArgSet("-rpcallowip")) { // Default to loopback if not allowing external IPs
+    if (!(gArgs.IsArgSet("-rpcallowip") && gArgs.IsArgSet("-rpcbind"))) { // Default to loopback if not allowing external IPs
```
Comment (kallewoof) [inline src/httpserver.cpp]: In commit `net: Always default rpcbind to localhost, never "all interfaces"`:

Nit: I think `if !a || !b` is easier on the eyes than `if !(a && b)`.

**Review by kallewoof — COMMENTED** [review]
utACK e0520135795563f13feedb28bad5aa3a7e9bc15d

**Comment by practicalswift** [comment]
Needs rebase :-)

**Comment by MarcoFalke** [comment]
@luke-jr This has had some review, but need rebase to be eligible for merge.

**Comment by luke-jr** [comment]
Rebased

**Comment by practicalswift** [comment]
utACK 27c44ef9c61f64d941ab82ec232a68141a2fde90

**Comment by MeshCollider** [comment]
re-utACK https://github.com/bitcoin/bitcoin/pull/14532/commits/27c44ef9c61f64d941ab82ec232a68141a2fde90

**Comment by laanwj** [comment]
Removing "needs release note" we have

from: https://github.com/bitcoin-core/bitcoin-devwiki/wiki/0.18.0-Release-Notes-Draft

> The `rpcallowip` option can no longer be used to automatically listen
  on all network interfaces.  Instead, the `rpcbind` parameter must also
  be used to specify the IP addresses to listen on.  Listening for RPC
  commands over a public network connection is insecure and should be
  disabled, so a warning is now printed if a user selects such a
  configuration.  If you need to expose RPC in order to use a tool
  like Docker, ensure you only bind RPC to your localhost, e.g. `docker
  run [...] -p 127.0.0.1:8332:8332` (this is an extra `:8332` over the
  normal Docker port specification).

**Pull request merged**